pytorch_lightning/utilities/apply_func.py

```diff
@@ -222,6 +222,53 @@ def apply_to_collections(
         ]
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if _is_dataclass_instance(data1):
+        assert _is_dataclass_instance(data2), (
+            "Expected inputs to be dataclasses of the same type or to have identical fields"
+            f" but got input 1 of type {type(data1)} and input 2 of type {type(data2)}"
+        )
```

Comment:
assertions are not meant for user input validation, let's raise an error instead :)

```suggestion
        if not _is_dataclass_instance(data2):
            raise TypeError(
                "Expected inputs to be dataclasses of the same type or to have identical fields"
                f" but got input 1 of type {type(data1)} and input 2 of type {type(data2)}"
            )
```